test/common/common/numeric_test.cc

```diff
@@ -0,0 +1,82 @@
+#include <functional>
+#include <limits>
+#include <random>
+
+#include "common/common/numeric.h"
+
+#include "test/test_common/test_base.h"
+
+namespace Envoy {
+
+std::random_device rd;
+std::mt19937 gen(rd());
+std::uniform_int_distribution<int32_t> negative_ints(std::numeric_limits<int32_t>::min(), -1);
+std::uniform_int_distribution<int32_t> nonnegative_ints(0, std::numeric_limits<int32_t>::max());
+std::uniform_int_distribution<int32_t> signed_ints(std::numeric_limits<int32_t>::min(),
+                                                   std::numeric_limits<int32_t>::max());
+std::uniform_int_distribution<uint32_t> unsigned_ints(0, std::numeric_limits<uint32_t>::max());
+
+template <typename T>
+void law(std::uniform_int_distribution<T>& distribution, std::function<void(T)> f) {
+  for (int i = 0; i < 100; ++i) {
+    f(distribution(gen));
+  }
+}
+
+template <typename T>
+void law(std::uniform_int_distribution<T>& distribution, std::function<void(T, T)> f) {
+  for (int i = 0; i < 100; ++i) {
+    f(distribution(gen), distribution(gen));
+  }
+}
+
+template <typename T>
+void law(std::uniform_int_distribution<T>& distribution, std::function<void(T, T, T)> f) {
+  for (int i = 0; i < 100; ++i) {
+    f(distribution(gen), distribution(gen), distribution(gen));
+  }
+}
+
+TEST(NumericTest, Abs) {
+  law<int32_t>(negative_ints, [](int32_t n) { EXPECT_EQ(-n, abs(n)); });
+  law<int32_t>(nonnegative_ints, [](int32_t n) { EXPECT_EQ(n, abs(n)); });
+  law<uint32_t>(unsigned_ints, [](uint32_t n) { EXPECT_EQ(n, abs(n)); });
+}
+
+TEST(NumericTest, Gcd) {
+  // some simple concrete examples
+  EXPECT_EQ(1, gcd(5, 7));
+  EXPECT_EQ(2, gcd(2, 4));
+
+  // definition: gcd(0, n) === gcd(n, 0) === abs(n)
+  law<int32_t>(signed_ints, [](int32_t n) {
+    EXPECT_EQ(abs(n), gcd(0, n));
+    EXPECT_EQ(abs(n), gcd(n, 0));
+  });
+  law<uint32_t>(unsigned_ints, [](uint32_t n) {
+    EXPECT_EQ(n, gcd(0U, n));
+    EXPECT_EQ(n, gcd(n, 0U));
+  });
+
+  // euclid's: gcd(a, b) === gcd(b, abs(b) % abs(a))
```

Comment:
This feels like we're rewriting Haskell Quickcheck in C++ with templates and HOFs :) It's pretty cool, but don't we just need like 4 or 5  test cases for something like abs or gcd?

Reply:
I'm a big fan of QuickCheck and property-based testing in certain situations, and this feels like one of those situations. Or, to put it another way, if I'm dumb enough to implement my own `abs` and `gcd`, I'm definitely dumb enough to test them this way.

The honest truth, though, is that I started out writing the tests as individual cases, and started feeling a combination of nervous about what cases I was forgetting and weirded-out by how arbitrary my cases seemed.

Reply:
We could just on-board as a dependency something like https://github.com/emil-e/rapidcheck. I don't think this was a big thing last time we discussed this topic. Doesn't have to be in this PR, but it does the real Qucikcheck thing of test case minimization. seems to have gtest/gmock integration too.

Reply:
I've put that on my to-do list, but I'll leave it alone for this PR.